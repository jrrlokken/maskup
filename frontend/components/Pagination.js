import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';
import Head from 'next/head';
import Link from 'next/link';

import PaginationStyles from './styles/PaginationStyles';
import DisplayError from './ErrorMessage';
import { perPage } from '../config';

export const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    _allProductsMeta {
      count
    }
  }
`;

const Pagination = ({ page }) => {
  const { data, error, loading } = useQuery(PAGINATION_QUERY);
  if (error) return <DisplayError error={error} />;
  if (loading) return <p>Loading...</p>;
  const { count } = data._allProductsMeta;
  const pageCount = Math.ceil(count / perPage);

  return (
    <PaginationStyles>
      <Head>
        <title>Mask Up! - Page {page} of {pageCount}</title>
      </Head>
      <Link href={`/products/${page - 1}`}>
        <a aria-disabled={page <= 1}>
          ↩ Prev
        </a>
      </Link>
      <p>
        Page {page} of {pageCount}
      </p>
      <p>{count} Items Total</p>
      <Link href={`/products/${page + 1}`}>
        <a aria-disabled={page >= pageCount}>
          Next ↪ 
        </a>
      </Link>
    </PaginationStyles>
  );
}

export default Pagination;