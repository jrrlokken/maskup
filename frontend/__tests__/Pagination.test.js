import { configure, mount } from 'enzyme';
import Adapter from '@hteker/enzyme-adapter-react-17';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from '@apollo/client/testing';
import { act } from 'react-dom/test-utils';
import wait from 'waait';

import Pagination, { PAGINATION_QUERY } from '../components/Pagination';

configure({ adapter: new Adapter() });

function makeMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: { 
        data: {
          _allProductsMeta: {
            count: length
          }
        }
      }
    }
  ];
}

describe('<Pagination/>', () => {
  it('displays a loading message', () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(1)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    const pagination = wrapper.find('Pagination');
    expect(toJSON(pagination)).toMatchSnapshot();
    expect(pagination.text()).toContain('Loading...');
  });
  it('renders pagination for 18 items', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await act(async () => {
      await wait();
      wrapper.update();
    
      expect(wrapper.text()).toContain('18 Items Total');
      expect(wrapper.text()).toContain('Page 1 of 5');
      expect(wrapper.debug()).toMatchSnapshot();
    });
  });
  it('disables prev button on first page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await act(async () => {
      await wait();
      wrapper.update();
      const prevButton = wrapper.find('a').first();
      expect(prevButton.prop('aria-disabled')).toEqual(true);
    });
  });
  it('disables next button on last page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={5} />
      </MockedProvider>
    );
    await act(async () => {
      await wait();
      wrapper.update();
      const nextButton = wrapper.find('a').at(1);
      expect(nextButton.prop('aria-disabled')).toEqual(true);
    });
  });
  it('enables both buttons on middle pages', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={3} />
      </MockedProvider>
    );
    await act(async () => {
      await wait();
      wrapper.update();
      const prevButton = wrapper.find('a').at(0);
      const nextButton = wrapper.find('a').at(1);
      expect(prevButton.prop('aria-disabled')).toEqual(false);
      expect(nextButton.prop('aria-disabled')).toEqual(false);
    });
  })
});
