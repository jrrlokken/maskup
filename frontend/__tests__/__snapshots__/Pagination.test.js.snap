// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> displays a loading message 1`] = `
<Pagination
  page={1}
>
  <p>
    Loading...
  </p>
</Pagination>
`;

exports[`<Pagination/> renders pagination for 18 items 1`] = `
"<MockedProvider mocks={{...}} addTypename={true}>
  <ApolloProvider client={{...}}>
    <Pagination page={1}>
      <PaginationStyles>
        <div className=\\"PaginationStyles-sc-15b0bt5-0 itHVNY\\">
          <Head>
            <_default reduceComponentsToState={[Function: reduceComponents]} headManager={{...}} inAmpMode={false} />
          </Head>
          <Link href=\\"/products/0\\">
            <a aria-disabled={true} onClick={[Function: onClick]} onMouseEnter={[Function (anonymous)]} href=\\"/products/0\\">
              ↩ Prev
            </a>
          </Link>
          <p>
            Page 
            1
             of 
            5
          </p>
          <p>
            18
             Items Total
          </p>
          <Link href=\\"/products/2\\">
            <a aria-disabled={false} onClick={[Function: onClick]} onMouseEnter={[Function (anonymous)]} href=\\"/products/2\\">
              Next ↪
            </a>
          </Link>
        </div>
      </PaginationStyles>
    </Pagination>
  </ApolloProvider>
</MockedProvider>"
`;
